File: migrate/tag.go
package migrate

import (
	"github.com/axumrs/axum-rs-migration/db"
	"github.com/axumrs/axum-rs-migration/model"
)

type Tag struct {
	Truncate bool
}

func (t *Tag) Read() (ls []*model.PGTag, err error) {
	if err := db.PG.Select(&ls, "SELECT * FROM tag ORDER BY id ASC"); err != nil {
		return nil, err
	}
	return ls, nil
}

func (t *Tag) Write(ls []*model.MySQLTag) error {
	tx, err := db.MY.Beginx()
	if err != nil {
		return err
	}

	if t.Truncate {
		if _, err := tx.Exec("TRUNCATE TABLE tag"); err != nil {
			return tx.Rollback()
		}
	}

	if _, err := tx.NamedExec("INSERT INTO tag (id,name, is_del) VALUES(:id,:name, :is_del)", ls); err != nil {
		return tx.Rollback()
	}

	return tx.Commit()
}

func (t *Tag) Migrate() error {
	ls, err := t.Read()
	if err != nil {
		return err
	}
	return t.Write(model.MySQLTagListFrom(ls))
}

File: migrate/tag.py
import db.pg as pg
import db.mysql as mysql
from utils import xid

def migration(cfg,truncate=True):
    rows = list()

    with mysql.conn(cfg, ) as c:
        with c.cursor(dictionary=True) as cur:
            cur.execute('SELECT id, name, is_del FROM tag ORDER BY id ASC')
            r = cur.fetchall()
            for x in r:
                rows.append(x)
    
    rows = tuple(map(lambda x: (xid.new(), x['name'], x['is_del']==1), rows))

    with pg.conn(cfg) as c:
        with c.cursor() as cur:
            if truncate:
                cur.execute('TRUNCATE TABLE tags')
            
            for row in rows:
                cur.execute('INSERT INTO tags (id, "name", is_del) VALUES(%s,%s,%s)', row)
    
    return len(rows)

def get_name(cfg, mysql_id):
    with mysql.conn(cfg, ) as c:
        with c.cursor(dictionary=True) as cur:
            cur.execute('SELECT name FROM tag WHERE id = %s', (mysql_id, ))
            r = cur.fetchone()
            if r:
                return r['name']
            else:
                return None

def get_id(cfg, name):
    with pg.conn(cfg) as c:
        with c.cursor() as cur:
            cur.execute('SELECT id FROM tags WHERE "name" = %s', (name, ))
            r = cur.fetchone()
            if r:
                return r['id']
            else:
                return None